plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/external/ProjectStateChangeListener.java

```diff
@@ -24,85 +24,88 @@
 
 /**
  * Resource change listener implementation for listening project open/close event and running the
- * {@link ExternalLibraryBuilderHelper external library build helper} according to the changes. Also got notified if new
+ * {@link ExternalLibraryBuilder external library build helper} according to the changes. Also got notified if new
  * workspace project is being created or an existing, accessible project is being deleted.
  */
 public class ProjectStateChangeListener implements IResourceChangeListener {
-
-	private static final Logger LOGGER = Logger.getLogger(ProjectStateChangeListener.class);
+	private static final Logger logger = Logger.getLogger(ProjectStateChangeListener.class);
 
 	@Inject
-	private ExternalLibraryWorkspace externalLibraryWorkspace;
+	private ExternalProjectProvider projectProvider;
 
 	@Inject
 	private ExternalLibraryBuildJobProvider buildJobProvider;
 
+	final private Collection<N4JSExternalProject> toClean = newLinkedHashSet();
+	final private Collection<N4JSExternalProject> toBuild = newLinkedHashSet();
+
 	@Override
 	public void resourceChanged(final IResourceChangeEvent event) {
-
 		if (null == event || null == event.getDelta()) {
 			return;
 		}
 
-		final Collection<IProject> toClean = newLinkedHashSet();
-		final Collection<IProject> toBuild = newLinkedHashSet();
-
 		try {
-			event.getDelta().accept(delta -> {
-
-				final IResource resource = delta.getResource();
-				if (resource instanceof IProject) {
+			toClean.clear();
+			toBuild.clear();
+			event.getDelta().accept(this::visit); // fill toClean and toBuild
 
-					final IProject project = (IProject) resource;
-					// XXX What if project is not accessible and/or not Xtext project with the corresponding builder?
-					final IProject externalProject = externalLibraryWorkspace.getProject(project.getName());
-					if (null != externalProject && externalProject.exists()) {
+			if (!toClean.isEmpty() || !toBuild.isEmpty()) {
+				logger.info("Received project open/close change.");
+				logger.info("Opened projects: " + Iterables.toString(from(toClean).transform(p -> p.getName())));
+				logger.info("Closed projects: " + Iterables.toString(from(toBuild).transform(p -> p.getName())));
 
-						if (CHANGED == delta.getKind() && (delta.getFlags() & OPEN) != 0) {
+				buildJobProvider.createBuildJob(toBuild, toClean).schedule();
+			}
 
-							// Workspace project close/open
-							if (project.isOpen()) {
-								toClean.add(externalProject);
-							} else {
-								toBuild.add(externalProject);
-							}
+		} catch (final CoreException e) {
+			logger.error("Error while visiting resource change event.", e);
+		}
+	}
 
-						} else if (REMOVED == delta.getKind()) {
+	boolean visit(IResourceDelta delta) {
+		IResource resource = delta.getResource();
+		if (resource instanceof IProject) {
 
-							// Workspace project deletion
-							toBuild.add(externalProject);
+			IProject project = (IProject) resource;
+			String name = project.getName();
```

Comment:
regarding the comment that was removed, check if `project.getName()` will not give you non-n4js project (e.g. just folder with `.project` file that eclipse recognises as `IProject`. If this can happen in this particular code path, maybe check if project contains manifest file.